packages/keystone/fields/CustomFile/Implementation.js

```diff
@@ -0,0 +1,184 @@
+              mimetype: String
+              encoding: String
+              publicUrl: String
+              meta: JSON
+            }
+          `,
+        ]
+    }
+
+    gqlOutputFieldResolvers () {
+        return {
+            [this.path]: (item, args, context) => {
+                const itemValues = item[this.path]
+                if (!itemValues) {
+                    return null
+                }
+
+                if (typeof itemValues === 'string') {
+                    return itemValues
+                }
+
+                const user = context?.req?.user || context?.authedItem || null
+
+                return {
+                    publicUrl: this.fileAdapter.publicUrl(itemValues, user),
+                    ...itemValues,
+                }
+            },
+        }
+    }
+
+    async validateInput (props) {
+        const { context, resolvedData } = props
+
+        const fileData = resolvedData[this.path]
+        if (fileData && fileData['meta']) {
+            if (fileData['meta']['authedItem'] !== context.authedItem.id) {
+                throw new GQLError({
+                    code: 'BAD_USER_INPUT',
+                    type: 'WRONG_FILE_ID',
+                    variable: [this.path],
+                    message: 'File not found or you do not have access to it',
+
+                }, context)
+            }
+        }
+
+        return await super.validateInput(props)
+    }
+
+    async resolveInput ({ resolvedData, existingItem, context, listKey }) {
+        const uploadData = resolvedData[this.path]
+        // New way to 'upload' - connect file
+        if (get(uploadData, 'signature')) {
+
+            let file
+            try {
+                file = jwt.verify(uploadData['signature'], this._fileSecret)
+            } catch (e) {
+                throw new GQLError({
+                    code: 'BAD_USER_INPUT',
+                    type: 'WRONG_SIGNATURE',
+                    variable: [this.path],
+                    message: 'Signature is wrong or expired',
+                })
+            }
+
+            const fileMeta = file.meta
+
+            if (fileMeta.authedItem !== context.authedItem.id) {
+                throw new GQLError({
+                    code: 'FORBIDDEN',
+                    type: 'ACCESS_DENIED',
+                    variable: [this.path],
+                    message: 'You are not own this file',
+                })
+            }
+
+            if (fileMeta.modelNames.length && !fileMeta.modelNames.includes(listKey)) {
+                throw new GQLError({
+                    code: 'FORBIDDEN',
+                    type: 'ACCESS_DENIED',
+                    variable: [this.path],
+                    message: 'Owner of this file restrict connection to this model',
+                })
+            }
+
+            let condoFile
+
+            if (this.fileServiceUrl) {
+                const response = await fetch(this.fileServiceUrl + '/api/files/attach', {
```

Comment:
We need to think about it, maybe task or remove?

Reply:
I think at the moment it should be removed, because it is not obvious who will use it except for its own developers